In [1]:
from datetime import datetime

In [2]:
import pickle

with open('coding.data', 'rb') as filehandle:
    # read the data as binary data stream
    placesList2 = pickle.load(filehandle)

In [3]:
codinglist=[]
for i in range(len(placesList2)):
    codinglist.append(placesList2[i])


In [4]:
codinglist

[[],
 [],
 [],
 [],
 [8209,
  131361,
  66113,
  1153,
  16658,
  8738,
  132162,
  67714,
  33300,
  17444,
  10308,
  135300,
  1048,
  34856,
  20552,
  8328],
 [131105,
  8389697,
  4196481,
  2101505,
  8705,
  263202,
  133186,
  8392834,
  4202754,
  2114050,
  526372,
  266308,
  139396,
  8405252,
  4227588,
  1052712,
  532552,
  278664,
  164104,
  8454664,
  8240,
  1065040,
  557200,
  327952,
  131600],
 [2097217,
  536875137,
  268443905,
  134234625,
  67142657,
  67585,
  4198466,
  2105474,
  536887554,
  268468738,
  134284290,
  67241986,
  8396868,
  4210820,
  2130180,
  536936964,
  268567556,
  134481924,
  16793672,
  8421512,
  4260104,
  2228744,
  537134088,
  268961800,
  33587280,
  16842896,
  8519952,
  4456976,
  2622480,
  537921552,
  65632,
  33685664,
  17039648,
  8913440,
  5243936,
  2099232],
 [33554561,
  34359755009,
  17179902465,
  8590001153,
  4295100417,
  2147749889,
  532481,
  67125378,
  33587458,
  34359804418,
  17180001282,
  85901

In [5]:
def mapcoding(n):
    l=codinglist[n]
    lu=codinglist[n+1]
    ln=[]
    co=-1
    li=0
    for i in range(len(l)):
        co=co+1
        if co ==n:
            co=0
            li=li+1
        
        ln.append(lu[(i+li)])
    return ln

In [6]:
import math
def puiss(i):
    p=int(math.pow(2,i))
    
    return p

In [7]:
def MAX(sets):
    #this function return max in integer set
    return (max(sets)) 

In [8]:
def settovector(ens,refvec):
    #this function convert set to vector
    vl=[]
    for i in range(len(refvec)):
           vl.append(0) 
    for i in ens:
        if i in refvec :
            vl[refvec.index(i)]=1
       
    return vl

In [9]:
def Bitset(i):
    #this function convert vector bit to integer
    y= ''.join(map(str, i)) 
    a=bin(int(y, 2))
    b=int(a, 2)
    return b

In [10]:
def liste2listint(l1,refvec):
    l2=[]
    for i in l1:
        l2.append(Bitset(settovector(i,refvec)))
    return l2

In [11]:
def chartoint(c):
    #this function convert character '1' to 1 and '0' to 0
    if c=='1':
        return 1
    else:
        return 0

In [12]:
def Convert(string): 
    #this function convert string to a list
    list1=[] 
    list1[:0]=string 
    return list1

In [13]:
def listchartoint(l):
    #this function convert lsit of char to bit vector
    l2=[]
    for i in l:
        l2.append(chartoint(i))
    return l2

In [14]:
def vectortoset(v,refvec):
    #this function convert binary vector to set 
    s=[]
    for i in  range(len(v)) :
        if v[i]==1:
            s.append(refvec[i])
    return set(s)

In [15]:
def decimalToBinary(n): 
    #this function convert integer to binary string
    return bin(n).replace("0b", "") 

In [16]:
def remplir(lm,c):
    #this function fill the missing 0 in vector
    k=[]
    f=c-len(lm)
    for i in range(f):
        k.append(0)
    k=k+lm
    return k

In [17]:
def inttoset(n,refvec):
    #this function convert integer to set
    
        f=decimalToBinary(n)
        #l=Convert(f)
        l3=listchartoint(f)
        l4=remplir(l3,len(refvec))
        return vectortoset(l4,refvec)
   

In [18]:
def listeint2listset(l,refvec):
    l2=[]
    for i in l or []:
        l2.append(inttoset(i,refvec))
    return l2

In [19]:
def oneinttobitvect(k,refvec):
    vec=[]
    for i in range(len(refvec)):
           vec.append(0) 
    if k in refvec:
        vec[refvec.index(k)]=1
    return Bitset(vec)
    

In [20]:
def symboliclique(s,refvec):
    c=Bitset(settovector(s,refvec))
    return c

In [21]:
def ensclique(c,refvec):
    s=set([])
    s=inttoset(c,refvec)
    return s

In [22]:
def splitt2(i,c,zero,one,treated):
        m=refvec[i]
        r=ensclique(c,refvec)
        s= list(r)
        #t=r-treated
        zero=list(filter(lambda x:   (x &m==0), s))
        zero.append(m)
        one=list(r-set([m]))
        #list(filter(lambda x:  ((x!=m) and  (x &m!=0)), s))
        mate=symboliclique(set(zero),refvec)
        rival=symboliclique(set(one),refvec)
        return (mate,rival)

In [23]:
def  countSetBits(n): 
    count = 0
    while (n): 
        count += n & 1
        n >>= 1
    return count 

In [24]:
with open('n.data', 'rb') as filehandle:
    # store the data as binary data stream
    N=pickle.load(filehandle)

In [25]:
n=N[0]

In [26]:
l=codinglist[n+1]

In [27]:
len(l)

25

In [28]:
Minit=[]
i=n
#print(i)
while i < (len(l)-1):
        Minit.append(l[i])
        i=i+(n+1)
#print(i)
i=i-(n+1)
i=i+1
while i <= (len(l)-1):
    Minit.append(l[i])
    i=i+1

In [29]:
Minit

[8705, 2114050, 4227588, 8454664, 8240, 1065040, 557200, 327952, 131600]

In [30]:
def OrOp(c,refvec):
    
    s=ensclique(c,refvec)
    r=0
    for e in s :
        r=r|e
    return r

In [31]:
%%time
#l=[3840, 61441, 458754, 528644, 74248, 1721360,1345568]  
refvec=l             
nn=len(Minit)
#nn=math.ceil(math.sqrt(n))
M=[symboliclique(set(Minit),refvec)]

mi=Minit[0]
#id=1086357584
#iid=refvec.index(id)
#idvec=oneinttobitvect(id,refvec)
ens=set([])      #[1086357584,67117216]
idvec=Bitset(settovector(ens,refvec))
#print(idvec)
i=0
treated=set([])
#print(" iteration N= ",i,"treaed =",treated)
#print("partition =",i,"=",refvec)
#s=[ensclique(e)for e in M]
#print("partition =",i,"=",s)
#print("-------------------------------------------------" )
while (len(treated) !=nn):
         if (mi not in treated) :
                treated.add(mi) 
                #print("treated=",treated)
                #print("*************************")
                temp=[]
                for c in M:
                    r=ensclique(c,refvec)
                    if mi in r :
                        z1=[]
                        o1=[]
                        (z1,o1)=splitt2(refvec.index(mi),c,z1,o1,treated)
                        #print(ensclique(z1))
                        #print(ensclique(o1)) 
                        if idvec&z1==idvec:
                               if z1 not in temp :
                                      if countSetBits(z1)>= 2:
                                            temp.append(z1)
                        if idvec&o1==idvec:
                               if o1 not in temp :
                                    if countSetBits(o1)>= 2:
                                          temp.append(o1)
                    else:
                         if idvec&c==idvec:
                            if c not in temp :
                                      temp.append(c)
                    
         else:
                i=i+1
                mi=Minit[i]      
                
         
         M= temp           #eliminationdouble1(temp)
         #print(" iteration N= ",i,"treaed =",treated)
         #print("partition",i, "=",M)
         #s=[ensclique(e)for e in M]
         #print(s)
         #print("-------------------------------------------------" )
    
    
#print(i)
#M=eliminationdouble1(temp)
Mi=list(set(M))    #eliminationdouble1(M)    
Tor={}
#TorN={}

co=0
for e in range(len(Mi)) :
     if countSetBits(Mi[e])==2:
             co=co+1
             print("symbolic representation ",Mi[e])
             Tor[Mi[e]]=OrOp(Mi[e],refvec)
             print("set representation ",ensclique(Mi[e],refvec))
             #print("*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*")
     

     
print("The number of solution is ",co)     

symbolic representation  1048578
set representation  {327952, 8705}
symbolic representation  32770
set representation  {327952, 2114050}
symbolic representation  1048580
set representation  {557200, 8705}
symbolic representation  32772
set representation  {557200, 2114050}
symbolic representation  1026
set representation  {327952, 4227588}
symbolic representation  36
set representation  {8454664, 557200}
symbolic representation  1048584
set representation  {1065040, 8705}
symbolic representation  1032
set representation  {1065040, 4227588}
symbolic representation  40
set representation  {8454664, 1065040}
symbolic representation  32784
set representation  {8240, 2114050}
symbolic representation  1040
set representation  {8240, 4227588}
symbolic representation  48
set representation  {8454664, 8240}
The number of solution is  12
Wall time: 15 ms


In [32]:
Tor

{1048578: 336657,
 32770: 2442002,
 1048580: 565905,
 32772: 2671250,
 1026: 4555540,
 36: 9011864,
 1048584: 1073745,
 1032: 5292628,
 40: 9519704,
 32784: 2122290,
 1040: 4235828,
 48: 8462904}

In [33]:
len(Tor)

12

In [34]:
bb=(n)*(n-1)
print(bb)

12


In [35]:
import pickle

# define a list of places


with open('TorNMinus2Key.data', 'rb') as filehandle:
    # store the data as binary data stream
    TorNMinus2Key=pickle.load(filehandle)

In [36]:
 TorNMinus2Key

[33280,
 33024,
 32770,
 32772,
 16640,
 16512,
 16392,
 16386,
 32784,
 8208,
 8196,
 4100,
 640,
 4224,
 513,
 16416,
 5120,
 4128,
 2080,
 516,
 8193,
 8200,
 40,
 1026,
 16385,
 264,
 129,
 4098,
 32832,
 10240,
 8256,
 6144,
 2049,
 384,
 1032,
 65,
 320,
 130,
 2052,
 2064,
 258,
 20,
 1040,
 24]

In [37]:
len(TorNMinus2Key)

44

In [38]:
import pickle

# define a list of places


with open('TorNMinus1Key.data', 'rb') as filehandle:
    # store the data as binary data stream
    TorNMinus1Key=pickle.load(filehandle)

In [39]:
TorNMinus1Key

[33026,
 33284,
 16768,
 16648,
 16642,
 16513,
 16514,
 10256,
 4226,
 32788,
 8212,
 8216,
 2068,
 1048,
 641,
 6176,
 16424,
 5122,
 33088,
 8257,
 10241,
 10244,
 6148,
 386]

In [40]:
import pickle

# define a list of places


with open('TorNKey.data', 'rb') as filehandle:
    # store the data as binary data stream
    TorNKey=pickle.load(filehandle)

In [41]:
TorNKey

[16770, 10260]

In [42]:
nextL=mapcoding(n)
refvec1=codinglist[n]
refvec2=codinglist[n+1]
TorNMinus2updated={}
for k in TorNMinus2Key :
          s1=ensclique(k,refvec1)
          s2=set([])
          r=0
          for e in s1 :
                   e1=nextL[refvec1.index(e)]
                   s2.add(e1)
                   r=r|e1
          k2=symboliclique(s2,refvec2)
          TorNMinus2updated[k2]=r

In [43]:
nextL=mapcoding(n)
refvec1=codinglist[n]
refvec2=codinglist[n+1]
TorNMinus1updated={}
for k in TorNMinus1Key :
          s1=ensclique(k,refvec1)
          s2=set([])
          r=0
          for e in s1 :
                   e1=nextL[refvec1.index(e)]
                   s2.add(e1)
                   r=r|e1
          k2=symboliclique(s2,refvec2)
          TorNMinus1updated[k2]=r

In [44]:
TorNMinus1updated

{16842880: 4612523,
 16908544: 9056491,
 8470528: 13118823,
 8454656: 13645163,
 8454272: 12871115,
 8405056: 9080173,
 8405120: 9194733,
 4720640: 12864935,
 2113664: 2906541,
 16779520: 9068909,
 4196608: 13134285,
 4196864: 13654445,
 526592: 9201006,
 264704: 9591150,
 147520: 9083310,
 2625536: 2504103,
 8393216: 9581805,
 2359424: 2513355,
 16850944: 4600167,
 4202560: 4626893,
 4718656: 4623787,
 4718848: 4992235,
 2621696: 2897259,
 82048: 5007790}

In [45]:
TorNMinus1updated.keys()

dict_keys([16842880, 16908544, 8470528, 8454656, 8454272, 8405056, 8405120, 4720640, 2113664, 16779520, 4196608, 4196864, 526592, 264704, 147520, 2625536, 8393216, 2359424, 16850944, 4202560, 4718656, 4718848, 2621696, 82048])

In [46]:
nextL=mapcoding(n)
refvec1=codinglist[n]
refvec2=codinglist[n+1]
TorNupdated={}
for k in TorNKey :
          s1=ensclique(k,refvec1)
          s2=set([])
          r=0
          for e in s1 :
                   e1=nextL[refvec1.index(e)]
                   s2.add(e1)
                   r=r|e1
          k2=symboliclique(s2,refvec2)
          TorNupdated[k2]=r

In [47]:
TorNupdated

{8470656: 13397487, 4720896: 13397487}

In [48]:
sol=[]
sol3=[]

In [49]:
start = datetime.now()
sol2=[]
avoid4={}
for e in Minit :
    if e !=Minit[len(Minit)-1]:
        for k,v in TorNupdated.items() :
               b= e & v
               if  b == 0 :
                             s1=ensclique(k,refvec2)
                             s1.add(e)
                             sol2.append(symboliclique(s1,refvec2))
                             if k in avoid4.keys() :
                                    avoid4[k].union(s1)
                             else:
                                     avoid4[k]=s1
               else:
                       for k1,v1 in Tor.items() :
                            s2=ensclique(k1,refvec2)
                            if e in s2:
                                 s1=ensclique(k,refvec2)
                                 for d in s1 :
                                         if d & v1 ==0 :
                                               s2.add(d)
                                 c=symboliclique(s2,refvec2)
                                 if len(s2)==n:  
                                              if c not in sol3 :
                                                        sol3.append(c)
                                 else:
                    
                                                 if len(s2)==n-1:
                                                         if c not in sol:
                                                                sol.append(c)
    else:
        
               e=Minit[len(Minit)-1]
               for k,v in TorNupdated.items() :
                    b= e & v
                    if  b == 0 :
                             s1=ensclique(k,refvec2)
                             s1.add(e)
                             sol2.append(symboliclique(s1,refvec2))
                             if k in avoid4.keys() :
                                    avoid4[k].union(s1)
                             else:
                                     avoid4[k]=s1
                    else:
                        
                                s1=ensclique(k,refvec2)
                                s=set([e])
                                for  d in s1:
                                    if d&e==0:
                                         s.add(d)
                                
                                c=symboliclique(s2,refvec2)
                                if len(s)==n:  
                                              if c not in sol3 :
                                                        sol3.append(c)
                                else:
                    
                                                 if len(s)==n-1:
                                                         if c not in sol:
                                                                sol.append(c)
                        
                        
                        
                        
end = datetime.now()
diff = end - start
print(diff.total_seconds())

0.002997


In [50]:
for e in sol3 :
    print(ensclique(e,refvec2))

{557200, 8705, 263202, 8405252}
{327952, 8389697, 2114050, 526372}
{532552, 4196481, 2114050, 327952}
{278664, 8389697, 8240, 4227588}
{8454664, 4202754, 1065040, 526372}
{8454664, 4196481, 263202, 1065040}


In [51]:
for e in sol :
    print(ensclique(e,refvec2))

{327952, 8705, 526372}
{1065040, 8705, 526372}
{1065040, 8705, 263202}
{557200, 8389697, 2114050}
{8240, 8389697, 2114050}
{8240, 4196481, 2114050}
{327952, 8389697, 4227588}
{532552, 327952, 4227588}
{1065040, 263202, 4227588}
{8454664, 4202754, 557200}
{8454664, 263202, 557200}
{8454664, 4196481, 8240}


In [52]:
sol2

[8470657, 4720897]

In [53]:
sol

[1064962,
 1064968,
 1572872,
 8421380,
 8421392,
 4227088,
 8389634,
 1282,
 525320,
 65572,
 524324,
 4194352]

In [54]:
sol3

[1574916, 8437762, 4227330, 8389776, 81960, 4718632]

In [55]:
print(n)

4


In [56]:
avoid4

{8470656: {131600, 278664, 526372, 4202754, 8389697},
 4720896: {131600, 263202, 532552, 4196481, 8405252}}

In [57]:
sol

[1064962,
 1064968,
 1572872,
 8421380,
 8421392,
 4227088,
 8389634,
 1282,
 525320,
 65572,
 524324,
 4194352]

In [58]:
for e in sol3 :
    print(ensclique(e,refvec2))

{557200, 8705, 263202, 8405252}
{327952, 8389697, 2114050, 526372}
{532552, 4196481, 2114050, 327952}
{278664, 8389697, 8240, 4227588}
{8454664, 4202754, 1065040, 526372}
{8454664, 4196481, 263202, 1065040}


In [59]:
sol3

[1574916, 8437762, 4227330, 8389776, 81960, 4718632]

In [60]:
sol2

[8470657, 4720897]

In [61]:
start = datetime.now()
sol1=[]
avoid={}
for k1,v1 in Tor.items() :
    for k,v in TorNMinus1updated.items() :
          if v1 & v == 0 :
                s1=ensclique(k,refvec2)
                #print("s1",s1)
                s2=ensclique(k1,refvec2)
                #print("s2",s2)
                for d in s2 :
                    s1.add(d)
                #print("s1",s1)
                c=symboliclique(s1,refvec2)
                #print(countSetBits(c))
                if len(s1)==n+1:
                          sol1.append(c)
                if k in avoid.keys() :
                     avoid[k].union(s2)
                else:
                    avoid[k]=s2
          else:
            
                       s1=ensclique(k,refvec2)
                       s2=ensclique(k1,refvec2)
                       for x in s2 :
                            s3=set([x])
                            for e in s1 :
                               if x & e== 0 :
                                   s3.add(e)
                       if len(s3)==n:
                             c=symboliclique(s3,refvec2)
                             if c not in sol3:
                                       sol3.append(c) 
                       else:
                                 
                                 if len(s3)==n-1:
                                          c=symboliclique(s3,refvec2)
                                          if c not in sol:
                                              sol.append(c) 
          
            
end = datetime.now()
diff = end - start
print(diff.total_seconds())           

0.016988


In [62]:
for e in sol3 :
    print(ensclique(e,refvec2))

{557200, 8705, 263202, 8405252}
{327952, 8389697, 2114050, 526372}
{532552, 4196481, 2114050, 327952}
{278664, 8389697, 8240, 4227588}
{8454664, 4202754, 1065040, 526372}
{8454664, 4196481, 263202, 1065040}
{1052712, 8705, 133186, 8405252}
{164104, 8705, 8392834, 526372}
{164104, 8389697, 2114050, 526372}
{164104, 4196481, 2114050, 266308}
{1052712, 8389697, 2114050, 139396}
{278664, 2101505, 133186, 4227588}
{532552, 2101505, 263202, 4227588}
{1052712, 8389697, 4202754, 557200}
{1052712, 133186, 8405252, 557200}
{532552, 131105, 8392834, 4227588}
{164104, 8392834, 1065040, 526372}
{164104, 4196481, 263202, 1065040}
{278664, 2101505, 133186, 8240}
{164104, 4196481, 8240, 266308}


In [63]:
%%time
for e in Minit :
    for k,v in TorNMinus1updated.items()  :
          b= e & v
          if  b == 0 :
                s1=ensclique(k,refvec2)
                s1.add(e)
                c=symboliclique(s1,refvec2)
                if len(s1)==n:  
                          if c not in sol3 :
                               sol3.append(c)
                else:
                    
                          if len(s1)==n-1:
                                if c not in sol:
                                     sol.append(c)
                
          else :
                   if k not in sol:
                         if k not in avoid.keys():
                               sol.append(c)

Wall time: 3.99 ms


  bo=True 
                    for y,r in avoid4.items():
                              if c & y== c :
                                  bo=False
                    if bo :            

In [64]:
#avoid

In [65]:
sol1

[8425986, 1313284, 16909570, 16850980, 1196104, 2625576, 4235344, 2360464]

In [66]:
for e in sol3 :
    print(ensclique(e,refvec2))

{557200, 8705, 263202, 8405252}
{327952, 8389697, 2114050, 526372}
{532552, 4196481, 2114050, 327952}
{278664, 8389697, 8240, 4227588}
{8454664, 4202754, 1065040, 526372}
{8454664, 4196481, 263202, 1065040}
{1052712, 8705, 133186, 8405252}
{164104, 8705, 8392834, 526372}
{164104, 8389697, 2114050, 526372}
{164104, 4196481, 2114050, 266308}
{1052712, 8389697, 2114050, 139396}
{278664, 2101505, 133186, 4227588}
{532552, 2101505, 263202, 4227588}
{1052712, 8389697, 4202754, 557200}
{1052712, 133186, 8405252, 557200}
{532552, 131105, 8392834, 4227588}
{164104, 8392834, 1065040, 526372}
{164104, 4196481, 263202, 1065040}
{278664, 2101505, 133186, 8240}
{164104, 4196481, 8240, 266308}
{8454664, 2101505, 263202, 139396}
{8454664, 131105, 4202754, 266308}
{1065040, 2101505, 263202, 139396}
{557200, 131105, 4202754, 266308}
{532552, 131105, 8392834, 327952}
{1052712, 8389697, 139396, 327952}
{131600, 8389697, 4202754, 526372}
{1052712, 8389697, 4202754, 131600}
{278664, 8389697, 4202754, 131600

In [67]:
len(sol3)

38

In [68]:
%%time
#avoid={}
for k1,v1 in Tor.items() :
    for k,v in TorNMinus2updated.items() :
          if v1 & v == 0 :
                s1=ensclique(k,refvec2)
                #print("s1",s1)
                s2=ensclique(k1,refvec2)
                #print("s2",s2)
                for d in s2 :
                    s1.add(d)
                #print("s1",s1)
                c=symboliclique(s1,refvec2)
                if len(s1)==n :
                    if c not in sol3 :
                        #boo=True
                        #for t,u in avoid.items() :
                            #if c & t ==c :
                                #boo=False
                            #if boo :
                                  sol3.append(c)
                        
          else :
                        s1=ensclique(k,refvec2)
                        #print("s1",s1)
                        s2=ensclique(k1,refvec2)
                        for x in s1 :
                            s3=set([x])
                            for e in s2:
                                if x & e== 0 :
                                    s3.add(e)
                            
                            if len(s3)==n-1:
                                          c=symboliclique(s3,refvec2)
                                          if c not in sol:
                                                 sol.append(c)

Wall time: 15 ms


In [69]:
sol

[1064962,
 1064968,
 1572872,
 8421380,
 8421392,
 4227088,
 8389634,
 1282,
 525320,
 65572,
 524324,
 4194352,
 1065024,
 1065088,
 1574912,
 1051136,
 1310848,
 1056832,
 1572928,
 16810240,
 8437760,
 8421888,
 4227328,
 4227584,
 49216,
 16818176,
 4227136,
 16778368,
 8390144,
 8389760,
 2098304,
 16778496,
 525568,
 263680,
 2622464,
 16842756,
 8454148,
 526340,
 16779268,
 2564,
 2621444,
 8389124,
 2359300,
 16842760,
 16908296,
 81928,
 16456,
 4718600,
 2113544,
 4194376,
 2621448,
 8388752,
 8388688,
 4196368,
 2097296,
 264208,
 131152,
 4194384,
 4194384,
 4194384,
 4194384,
 4194384,
 4194384,
 4194384,
 4194384,
 4194384,
 4194384,
 4194384,
 4194384,
 4194384,
 1196096,
 1196096,
 1196096,
 1196096,
 1196096,
 1196096,
 1196096,
 1196096,
 8437824,
 8437824,
 8437824,
 8437824,
 8437824,
 8437824,
 8437824,
 4235328,
 4235328,
 4235328,
 4235328,
 4235328,
 16909568,
 16909568,
 16909568,
 16909568,
 16909568,
 16909568,
 16909568,
 16909568,
 16909568,
 16909568,
 16

In [70]:
sol3

[1574916,
 8437762,
 4227330,
 8389776,
 81960,
 4718632,
 1313280,
 1196096,
 8437824,
 4235328,
 8425984,
 2360448,
 2622720,
 8454660,
 264708,
 16909568,
 147528,
 4718664,
 2359440,
 4202576,
 2625568,
 16850976,
 2625544,
 16850948,
 16908546,
 8393218,
 8470529,
 8454657,
 8454273,
 8405121,
 4720641,
 2113665,
 4196609,
 4196865,
 526593,
 4718849,
 2621697,
 82049,
 1196034,
 1311234,
 16810242,
 8421890,
 8425474,
 4227586,
 37378,
 1311236,
 1312772,
 1051140,
 8421892,
 16818180,
 16909314,
 16778498,
 8390146,
 132354,
 263682,
 16842788,
 2359332,
 16785444,
 2621476,
 73764,
 1196040,
 1179720,
 1065032,
 1572936,
 16909320,
 2622472,
 16842792,
 2113576,
 2101288,
 528424,
 2621480,
 4227152,
 8421456,
 4235280,
 41040,
 2360336,
 263312,
 2098320,
 2359344,
 4202544]

In [71]:
len(sol3)

80

In [72]:
TorNKey=sol1+sol2

In [73]:
len(TorNKey)

10

In [74]:
for e in TorNKey :
    print(ensclique(e,refvec2))

{8389697, 2114050, 139396, 1052712, 327952}
{8705, 133186, 8405252, 1052712, 557200}
{131105, 8392834, 4227588, 532552, 327952}
{131105, 4202754, 266308, 8454664, 557200}
{8705, 8392834, 526372, 164104, 1065040}
{2101505, 263202, 139396, 8454664, 1065040}
{4196481, 2114050, 266308, 164104, 8240}
{2101505, 133186, 4227588, 278664, 8240}
{8389697, 4202754, 526372, 278664, 131600}
{4196481, 263202, 8405252, 532552, 131600}


In [75]:
len(set(TorNKey))

10

In [76]:
len(TorNKey)

10

In [77]:
import pickle

# define a list of places
pList = TorNKey

with open('TorNKey.data', 'wb') as filehandle:
    # store the data as binary data stream
    pickle.dump(pList, filehandle)

In [78]:

#sol=[]
#sol3=[]
for k1,v1 in Tor.items() :
    for k,v in TorNMinus2updated.items() :
          if v1 & v == 0 :
                s1=ensclique(k,refvec2)
                s2=ensclique(k1,refvec2)
                for d in s2 :
                    s1.add(d)
                c=symboliclique(s1,refvec2)
                #print(len(s1))
                if len(s1)==n:
                     if c not in sol3:
                         sol3.append(c)
                else:
                    
                      if len(s1)==n-1:
                            if c not in sol:
                                  sol.append(c)
          

In [79]:
#sol=[]
for e in Minit :
     for k,v in TorNMinus1updated.items() :
            #if k in avoid.keys():
                  #if e not in avoid[k] :
                        if e & v == 0 :
                             s1=ensclique(k,refvec2)
                             s1.add(e)
                             c=symboliclique(s1,refvec2)
                             if c not in sol3 :
                                    sol3.append(c)
                        else:
                            
                               s1=ensclique(k,refvec2)
                               for d in s1 :
                                   s=set([e])
                                   if e&d==0:
                                      s.add(d)
                               if len(s)==n-1:
                                   c=symboliclique(s,refvec2)
                                   if c not in sol :
                                         sol.append(c)
                            
            #else:
                        #if e & v == 0 :
                               #s1=ensclique(k,refvec2)
                               #s1.add(e)
                               #c=symboliclique(s1,refvec2)
                               #if c not in sol :
                                      #sol.append(c)

In [80]:
len(sol)

234

In [81]:
s=set(sol)
sol=list(s)

In [82]:
len(sol)

122

In [83]:
#sol3=[]
for e in Minit :
     for k,v in TorNMinus2updated.items() :
                        if e & v == 0 :
                             s1=ensclique(k,refvec2)
                             s1.add(e)
                             #print(s1)
                             c=symboliclique(s1,refvec2)
                             if len(s1)==n-1:
                                    if c not in sol:
                                        sol.append(c)
            

In [84]:
len(sol3)

80

In [85]:
sol3

[1574916,
 8437762,
 4227330,
 8389776,
 81960,
 4718632,
 1313280,
 1196096,
 8437824,
 4235328,
 8425984,
 2360448,
 2622720,
 8454660,
 264708,
 16909568,
 147528,
 4718664,
 2359440,
 4202576,
 2625568,
 16850976,
 2625544,
 16850948,
 16908546,
 8393218,
 8470529,
 8454657,
 8454273,
 8405121,
 4720641,
 2113665,
 4196609,
 4196865,
 526593,
 4718849,
 2621697,
 82049,
 1196034,
 1311234,
 16810242,
 8421890,
 8425474,
 4227586,
 37378,
 1311236,
 1312772,
 1051140,
 8421892,
 16818180,
 16909314,
 16778498,
 8390146,
 132354,
 263682,
 16842788,
 2359332,
 16785444,
 2621476,
 73764,
 1196040,
 1179720,
 1065032,
 1572936,
 16909320,
 2622472,
 16842792,
 2113576,
 2101288,
 528424,
 2621480,
 4227152,
 8421456,
 4235280,
 41040,
 2360336,
 263312,
 2098320,
 2359344,
 4202544]

for fg in TorNMinus1updated.keys() :
    if fg not in sol3 :
        sol3.append(fg)

soll3=[]
for c1 in sol :
    bbb=True
    for c2 in sol3 :
        if c1&c2==c1:
            bbb=False
            
    if bbb :
        soll3.append(c1)
sol3=soll3
        

In [86]:
len(sol3)

80

In [87]:
len(set(sol3))

80

ub=[]
for e in sol3 :
    s=ensclique(e,refvec2)
    u=set([])
    for z in s :
        u.add(refvec2.index(z))
    print(u)
    ub.append(u)
        

In [88]:
{24, 2, 11}
{2, 20, 13}
{24, 10, 7}
{20, 14, 7}
{24, 16, 7}
{18, 20, 7}
{0, 17, 8}
{24, 8, 11}
{17, 4, 6}
{18, 11, 4}
{11, 4, 22}
{17, 10, 4}
{0, 17, 14}
{20, 13, 6}
{16, 0, 13}
{0, 13, 22}

{0, 13, 22}

ub

jj=[{131600, 266308, 4196481},
 {8240, 4196481, 8405252},
 {131600, 526372, 8392834},
 {8240, 4227588, 8392834},
 {131600, 532552, 8392834},
 {8240, 164104, 8392834},
 {131105, 278664, 4202754},
 {131600, 266308, 4202754},
 {8705, 133186, 278664},
 {8705, 164104, 266308},
 {8705, 266308, 557200},
 {8705, 278664, 526372},
 {131105, 278664, 4227588},
 {8240, 133186, 8405252},
 {131105, 532552, 8405252},
 {131105, 557200, 8405252}]

sb=[]
for e in jj :
    u=set([])
    for z in e :
        u.add(refvec2.index(z))
    print(u)
    sb.append(u)

sb

for t in sb :
    if t not in ub:
        print(t)

In [89]:
for e in sol :
    print(ensclique(e,refvec2))

{8705, 263202, 8405252}
{1052712, 8705, 8405252}
{327952, 8705, 526372}
{327952, 8389697, 4227588}
{557200, 8389697, 2114050}
{557200, 131105, 4202754}
{1052712, 8389697, 2114050}
{1052712, 4196481, 2114050}
{1065040, 8705, 526372}
{1065040, 8705, 263202}
{1065040, 263202, 4227588}
{1065040, 131105, 4202754}
{1065040, 131105, 8392834}
{557200, 8389697, 4202754}
{1065040, 4202754, 526372}
{1065040, 4196481, 263202}
{8240, 8389697, 2114050}
{8240, 4196481, 2114050}
{8240, 4196481, 8405252}
{8240, 133186, 8405252}
{327952, 8705, 8392834}
{557200, 2101505, 263202}
{557200, 2101505, 133186}
{557200, 8705, 133186}
{557200, 8705, 263202}
{557200, 131105, 2114050}
{1052712, 8705, 327952}
{1052712, 327952, 2114050}
{8454664, 131105, 4202754, 266308}
{2101505, 263202, 4227588}
{8454664, 4202754, 557200}
{8454664, 263202, 557200}
{1052712, 8389697, 557200}
{1052712, 8389697, 4202754, 557200}
{1052712, 8705, 557200}
{1065040, 2101505, 526372}
{1065040, 2101505, 263202}
{1052712, 2114050, 557200}
{

In [90]:
len(sol)

196

In [91]:
soll=[]
for c1 in sol :
    bbb=True
    for c2 in TorNKey :
        if c1&c2==c1:
            bbb=False
            
    if bbb :
        soll.append(c1)
sol=soll

In [92]:
len(sol)

101

In [93]:
import pickle

# define a list of places
pList = sol

with open('TorNMinus2Key.data', 'wb') as filehandle:
    # store the data as binary data stream
    pickle.dump(pList, filehandle)

In [94]:
import pickle

# define a list of places
pList = sol3

with open('TorNMinus1Key.data', 'wb') as filehandle:
    # store the data as binary data stream
    pickle.dump(pList, filehandle)

In [95]:
nlist=[n+1]

In [96]:
import pickle

# define a list of places
ppList = nlist

with open('n.data', 'wb') as filehandle:
    # store the data as binary data stream
    pickle.dump(ppList, filehandle)